include/veriblock/pop/blockchain/base_block_tree.hpp

```diff
@@ -832,8 +897,6 @@ struct BaseBlockTree {
  protected:
   //! stores ALL blocks, including valid and invalid
   block_index_t blocks_;
-  //! stores all removed blocks, to ensure pointers to blocks remain stable
-  block_index_t removed_;
```

Comment:
is there any particular reason to do this?

Reply:
yes:
* less block hash recalculations. block hash is needed to move a block index to and from removed_. commands can resolve blocks early and operate on pointers instead. endorsements can point to blocks directly instead of storing hashes.
* tracking of blocks that are known to the altchain but don't belong to the best chain including VBK and BTC blocks, blocks in the mempool(temp block tree)

Reply:
okay, but why it is necessary to store these blocks in `blocks_`? 
Why not separate map?

Reply:
I'd rephrase that question as: whether a block deleted or not is just an one-bit flag, why store this bit of information in a separate map?
There are many disadvantages of having a separate map:

* deleting and restoring a block is more expensive as the hash has to be recalculated(or cached like what is currently done);
* searching for the corresponding blockindex across 2 maps is more expensive;
* 2 maps will consume more RAM for the same workload as it's expensive to shrink/grow hash maps, or alternatively tree-based maps will not provide amortized O(1) complexity;
* checking whether the block is deleted is expensive;
* data is best kept normalized, and that means it would be advantageous to go further and merge isDeleted flag and the reference counter such that isDeleted = (ref_counter == 0), as the alternative is more asserts, performance loss and potential for mistakes.